core/src/test/java/feign/AsyncFeignTest.java

```diff
@@ -0,0 +1,1067 @@
+
+		TestInterfaceAsync api = AsyncFeign.asyncBuilder().mapAndDecode(upperCaseResponseMapper(), new StringDecoder())
+				.target(TestInterfaceAsync.class, "http://localhost:" + server.getPort());
+
+		assertEquals("RESPONSE!", unwrap(api.post()));
+	}
+
+	@Test
+	public void beanQueryMapEncoderWithPrivateGetterIgnored() throws Exception {
+		TestInterfaceAsync api = new TestInterfaceAsyncBuilder().queryMapEndcoder(new BeanQueryMapEncoder())
+				.target("http://localhost:" + server.getPort());
+
+		PropertyPojo.ChildPojoClass propertyPojo = new PropertyPojo.ChildPojoClass();
+		propertyPojo.setPrivateGetterProperty("privateGetterProperty");
+		propertyPojo.setName("Name");
+		propertyPojo.setNumber(1);
+
+		server.enqueue(new MockResponse());
+		CompletableFuture<?> cf = api.queryMapPropertyPojo(propertyPojo);
+		assertThat(server.takeRequest()).hasQueryParams(Arrays.asList("name=Name", "number=1"));
+		checkCFCompletedSoon(cf);
+	}
+
+	@Test
+	public void queryMap_with_child_pojo() throws Exception {
+		TestInterfaceAsync api = new TestInterfaceAsyncBuilder().queryMapEndcoder(new FieldQueryMapEncoder())
+				.target("http://localhost:" + server.getPort());
+
+		ChildPojo childPojo = new ChildPojo();
+		childPojo.setChildPrivateProperty("first");
+		childPojo.setParentProtectedProperty("second");
+		childPojo.setParentPublicProperty("third");
+
+		server.enqueue(new MockResponse());
+		CompletableFuture<?> cf = api.queryMapPropertyInheritence(childPojo);
+		assertThat(server.takeRequest()).hasQueryParams("parentPublicProperty=third", "parentProtectedProperty=second",
+				"childPrivateProperty=first");
+		checkCFCompletedSoon(cf);
+	}
+
+	@Test
+	public void beanQueryMapEncoderWithNullValueIgnored() throws Exception {
+		TestInterfaceAsync api = new TestInterfaceAsyncBuilder().queryMapEndcoder(new BeanQueryMapEncoder())
+				.target("http://localhost:" + server.getPort());
+
+		PropertyPojo.ChildPojoClass propertyPojo = new PropertyPojo.ChildPojoClass();
+		propertyPojo.setName(null);
+		propertyPojo.setNumber(1);
+
+		server.enqueue(new MockResponse());
+		CompletableFuture<?> cf = api.queryMapPropertyPojo(propertyPojo);
+
+		assertThat(server.takeRequest()).hasQueryParams("number=1");
+
+		checkCFCompletedSoon(cf);
+	}
+
+	@Test
+	public void beanQueryMapEncoderWithEmptyParams() throws Exception {
+		TestInterfaceAsync api = new TestInterfaceAsyncBuilder().queryMapEndcoder(new BeanQueryMapEncoder())
+				.target("http://localhost:" + server.getPort());
+
+		PropertyPojo.ChildPojoClass propertyPojo = new PropertyPojo.ChildPojoClass();
+
+		server.enqueue(new MockResponse());
+		CompletableFuture<?> cf = api.queryMapPropertyPojo(propertyPojo);
+		assertThat(server.takeRequest()).hasQueryParams("/");
+
+		checkCFCompletedSoon(cf);
+	}
+
+	interface TestInterfaceAsync {
+
+		@RequestLine("POST /")
+		CompletableFuture<Response> response();
+
+		@RequestLine("POST /")
+		CompletableFuture<String> post() throws TestInterfaceException;
```

Comment:
It is possible, and easy to implement. The proxy ReflectiveAsyncFeign would need to detect these kinds of methods and use CompleteableFuture.join() in a try/catch block with appropriate exception handling.

I though would exercise some caution here.

1. I think that interfaces should either be asynchronous or synchronous - I really can't think of a case where it makes sense to mix between the two.
2. If this is to be used as a replacement for the synchronous approach today, then I think it does require adding in a "blessed" way for doing retries (as exists today). Note though that it would break existing integrations, which would need to implement AsyncClient, as well as client code, in particular because context is now explicit. One could get around this by having the client also implement the context supplier (and handle under the covers in the builder by checking if the client implements Supplier<C>). This isn't too bad, since the synchronous clients today handle this internally generally on a thread local anyway.

Reply:
> I think that interfaces should either be asynchronous or synchronous - I really can't think of a case where it makes sense to mix between the two.

You may not, but our user's may, so we should consider it.

See my other comment on Retry.  We will have to break it.

Reply:
There is really only one reason that I can see to want a mixed interface and that's when wants to extend an existing synchronous interface to add on new async features (and one wants to be able to pass the instance to existing code).

Otherwise, it's generally trivial to do a join on the async call. The other way, changing client code from sync to async would likely take additional effort. But you're right that users could potentially have other constraints or reasons they need things a certain way.

However, it's not difficult to support - best practices can certainly be captured in documentation.

Reply:
> I think that interfaces should either be asynchronous or synchronous - I really can't think of a case where it makes sense to mix between the two.

My point at supporting both at same time, would be mainly to force a single workflow inside feign.

I don't really expect anyone using async for 3 methods and having one not async.  Maybe someone will.

Reply:
I added now sync through async. One thing that was necessary as part of that was coercing the SynchronousMethodHandler to explictly call the decoder (which is really the stageDecode in AsyncFeign) - this is because there were some edge cases in terms of Response and void return types which have correct special handling in AsyncResponseHandler which would be quite ugly to handle otherwise. I thought this way would be the most robust, absent a more extensive rewrite. To keep things consistent and simple, the coercion is also used for the async case, even though it works without it.